### PR TITLE
[feature](function) add approx_top_k aggregation function

### PR DESCRIPTION
**Pull request by zzzxl1993**
## Proposed changes

1. select approx_top_k(clientip, status, size, 10, 300) from tbl;

      This code implements an approximate Top-N query function based on the SpaceSaving algorithm. SpaceSaving is an efficient streaming algorithm commonly used to handle frequent element query problems in large datasets. Below is a description of the main functionalities:
      
      (1) Data Structures and Memory Management:
      The SpaceSavingArena class provides a memory pool to manage memory allocation and deallocation. For keys of type StringRef, it handles the memory by copying the string into the memory pool.
      The Counter struct stores the key, count, and error for each element, and provides serialization and deserialization functions.

      (2) Insertion and Updates:
      The insert method is used to insert new elements or update the count of existing elements. If the current capacity is not full, it inserts the new element; if it is full, it replaces the element with the smallest count based on the element's count and error.

      (3) Merge Operation:
      The merge method allows merging two SpaceSaving objects. During the merge, it adjusts the counts and errors of the existing elements, ensuring that the result maintains the correct order.

      (4) Top-K Query:
      The top_k method returns the current Top-K most frequent elements, sorted by their count and error.

      (5) Capacity Expansion and Shrinking:
      The resize method allows adjusting the storage capacity, and it recalculates the size of the alpha_map accordingly.

      (6) Serialization and Deserialization:
      The write and read methods are provided for serializing the SpaceSaving structure to disk or reading data from disk.

      (7) Optimization and Performance:
      The code uses a hash table-based approach for lookup and storage, and dynamically adjusts the alpha_map size to optimize performance and reduce memory waste.

      In summary, the SpaceSaving class efficiently implements Top-N queries for large data streams within limited memory, with efficient insertion, updating, and merging mechanisms.

### CODE REVIEW

**Comment by zzzxl1993** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/catalog/BuiltinAggregateFunctions.java [hunk 1]
```diff
@@ -145,7 +146,8 @@ public class BuiltinAggregateFunctions implements FunctionHelper {
             agg(TopNWeighted.class, "topn_weighted"),
             agg(Variance.class, "var_pop", "variance_pop", "variance"),
             agg(VarianceSamp.class, "var_samp", "variance_samp"),
-            agg(WindowFunnel.class, "window_funnel")
+            agg(WindowFunnel.class, "window_funnel"),
+            agg(MultiTopN.class, "multi_topn")
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/BuiltinAggregateFunctions.java:150]: sort by lexicographical order

**Review by morrySnow — CHANGES_REQUESTED** [review]
add test case please

**Comment by morrySnow** [comment]
add desc, and add doc to https://github.com/apache/doris-website

**Comment by zzzxl1993** [comment]
run buildall

**Comment by zzzxl1993** [comment]
run buildall

**Comment by zzzxl1993** [comment]
run buildall

**Comment by airborne12** [comment]
need detail description for this PR

File: be/src/vec/aggregate_functions/aggregate_function_multi_top.h [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+};
+
+template <typename NestFunction, bool result_is_nullable>
+class AggregateFunctionNullVariadicInline;
+
+#define REGISTER_AGGREGATE_FUNCTION(Name) \
+    template <typename T>                 \
+    struct is_aggregate_function_##Name : std::is_base_of<AggregateFunctionMultiTop, T> {}
+
+REGISTER_AGGREGATE_FUNCTION(MultiTopN);
```
Comment (xiaokang) [inline be/src/vec/aggregate_functions/aggregate_function_multi_top.h:36]: use consistent lower case name style

File: be/src/vec/aggregate_functions/aggregate_function_multi_topn.h [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+private:
+    using State = AggregateFunctionTopKGenericData;
+
+public:
+    AggregateFunctionMultiTopN(const DataTypes& argument_types_)
+            : IAggregateFunctionDataHelper<AggregateFunctionTopKGenericData,
+                                           AggregateFunctionMultiTopN>(argument_types_),
+              column_size(argument_types_.size() - 2) {}
+
+    String get_name() const override { return "multi_topn"; }
```
Comment (xiaokang) [inline be/src/vec/aggregate_functions/aggregate_function_multi_topn.h:68]: approx_topn is better since the key semantics is approximate and multi filed is not necessary.

File: be/src/vec/aggregate_functions/aggregate_function_simple_factory.cpp [hunk 4]
```diff
@@ -114,20 +115,17 @@ AggregateFunctionSimpleFactory& AggregateFunctionSimpleFactory::instance() {
         register_aggregate_function_linear_histogram(instance);
         register_aggregate_function_map_agg(instance);
         register_aggregate_function_bitmap_agg(instance);
-
```
Comment (xiaokang) [inline be/src/vec/aggregate_functions/aggregate_function_simple_factory.cpp:117]: unnecessary change

File: be/src/vec/aggregate_functions/aggregate_function.h [hunk 5]
```diff
@@ -287,8 +288,14 @@ class IAggregateFunctionHelper : public IAggregateFunction {
     void add_batch_single_place(size_t batch_size, AggregateDataPtr place, const IColumn** columns,
                                 Arena* arena) const override {
         const Derived* derived = assert_cast<const Derived*>(this);
-        for (size_t i = 0; i < batch_size; ++i) {
-            derived->add(place, columns, i, arena);
+
+        if constexpr (is_aggregate_function_multi_top<Derived>::value ||
+                      is_aggregate_function_multi_top_with_null_variadic_inline<Derived>::value) {
+            derived->add_range(place, columns, 0, batch_size, arena);
```
Comment (xiaokang) [inline be/src/vec/aggregate_functions/aggregate_function.h:294]: add_range() calls add() in a loop just like the `else` branch. So is it necessary to add a branch for is_aggregate_function_multi_top?

```
    void add_range(AggregateDataPtr __restrict place, const IColumn** columns, ssize_t min,
                   ssize_t max, Arena* arena) const {
        for (ssize_t row_num = min; row_num < max; ++row_num) {
            add(place, columns, row_num, arena);
        }
    }
```

File: be/src/vec/aggregate_functions/aggregate_function_multi_topn.h [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+            arena->rollback(ref.size);
+        }
+
+        set.read_alpha_map(buf);
+
+        read_var_uint(threshold, buf);
+        read_var_uint(reserved, buf);
+    }
+
+    void add(AggregateDataPtr __restrict place, const IColumn** columns, ssize_t row_num,
```
Comment (xiaokang) [inline be/src/vec/aggregate_functions/aggregate_function_multi_topn.h:125]: add comment

File: be/src/vec/aggregate_functions/aggregate_function_multi_topn.h [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+    bool allocates_memory_in_arena() const override { return true; }
+
+    void serialize(ConstAggregateDataPtr __restrict place, BufferWritable& buf) const override {
+        this->data(place).value.write(buf);
+
+        write_var_uint(threshold, buf);
+        write_var_uint(reserved, buf);
+    }
+
+    void deserialize(AggregateDataPtr __restrict place, BufferReadable& buf,
```
Comment (xiaokang) [inline be/src/vec/aggregate_functions/aggregate_function_multi_topn.h:81]: add comment

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/agg/MultiTopN.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        this(distinct, false, varArgs);
+    }
+
+    public MultiTopN(boolean distinct, boolean alwaysNullable, Expression... varArgs) {
+        super("multi_topn", distinct, alwaysNullable, varArgs);
+    }
+
+    @Override
+    public void checkLegalityBeforeTypeCoercion() {
+        if (arity() < 3) {
```
Comment (xiaokang) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/agg/MultiTopN.java:58]: we should add a function without argument `reserved` and use a default value to be more easy to use for general users.

File: be/src/vec/aggregate_functions/aggregate_function.h (hunk 5 above)
Comment (zzzxl1993) [reply be/src/vec/aggregate_functions/aggregate_function.h:294]: Calling add directly on nullable types has poor performance. Using add_range uniformly can prevent multiple calls.

File: be/src/vec/aggregate_functions/aggregate_function.h [hunk 9]
```diff
@@ -287,8 +288,14 @@ class IAggregateFunctionHelper : public IAggregateFunction {
     void add_batch_single_place(size_t batch_size, AggregateDataPtr place, const IColumn** columns,
                                 Arena* arena) const override {
         const Derived* derived = assert_cast<const Derived*>(this);
-        for (size_t i = 0; i < batch_size; ++i) {
-            derived->add(place, columns, i, arena);
+
+        if constexpr (is_aggregate_function_multi_top<Derived>::value ||
```
[the remark this answers is not in the document]
Comment (zzzxl1993) [reply be/src/vec/aggregate_functions/aggregate_function.h:292]: Refer to the implementation in add_match

**Comment by zzzxl1993** [comment]
run buildall

**Comment by zzzxl1993** [comment]
run buildall

**Comment by zzzxl1993** [comment]
run buildall

**Comment by zzzxl1993** [comment]
run buildall

**Comment by zzzxl1993** [comment]
run buildall

File: be/src/vec/aggregate_functions/aggregate_function_approx_topn.h [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+                            _reserved);
+        }
+
+        _init_flag = true;
+    }
+
+    mutable size_t _column_size = 0;
+    mutable bool _init_flag = false;
+    mutable uint64_t _threshold = 10;
+    mutable uint64_t _reserved = 300;
```
Comment (zhangstar333) [inline be/src/vec/aggregate_functions/aggregate_function_approx_topn.h:283]: seems those variables could put in AggregateFunctionTopKGenericData,
and implement some function more clearly.

**Comment by zzzxl1993** [comment]
run buildall

**Comment by zzzxl1993** [comment]
run buildall

**Review by morrySnow — DISMISSED** [review]

File: be/src/vec/aggregate_functions/aggregate_function_approx_topn.h (hunk 10 above)
Comment (zzzxl1993) [reply be/src/vec/aggregate_functions/aggregate_function_approx_topn.h:276]: 1. remove the use of optimization for is_aggregate_function_multi_top_with_null_variadic_inline.
2. These parameters won’t be used in AggregateFunctionTopKGenericData, so it seems better to place them outside.

**Comment by zzzxl1993** [comment]
run buildall

**Review by xiaokang — APPROVED** [review]
LGTM

**Review by qidaye — APPROVED** [review]
LGTM
